File: validation/src/main/java/com/validation/controller/ApiController.java
package com.validation.controller;


import com.validation.dto.PersonRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.support.DefaultMessageSourceResolvable;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import javax.validation.ConstraintViolation;
import javax.validation.Valid;
import javax.validation.Validator;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Validated
@RestController
@RequestMapping(value = "/api")
public class ApiController {

    private final Logger LOGGER = LoggerFactory.getLogger(this.getClass());
//    @Autowired
//    private ApiService apiService;

    @Autowired
    Validator validator;

    @PostMapping("/validateAuto")
    public void validateAuto(@Valid @RequestBody PersonRequest request) {

    }

    @PostMapping("/validateByHand")
    public void validateByHand(@RequestBody PersonRequest request) {
        Set<ConstraintViolation<PersonRequest>> violations = validator.validate(request);
        for (ConstraintViolation<PersonRequest> v : violations) {
            System.out.println(v.getPropertyPath() + ": " + v.getMessage());
        }
    }

    @PostMapping("/dept/{empId}/{depthId}")
    public void updateEmpDepartment(
        @PathVariable("empId") @Min(value = 10, message = "empId >= 10") @Max(value = 100, message = "empId <= 100")
        Integer emp_id,

        @PathVariable("depthId") @Min(value = 1001, message = "depthId >= 1001")
        Integer depthId
    ) {

    }

    @PostMapping("/validateBinding")
    public <T> T validateBinding(@Valid @RequestBody PersonRequest request, BindingResult bindingResult) {
        if(bindingResult.hasErrors()) {
            Map<String, String> errorDetails = bindingResult.getFieldErrors().stream()
                .collect(Collectors.toMap(FieldError::getField, DefaultMessageSourceResolvable::getDefaultMessage));
            return (T) errorDetails;
        }
        return null;
    }

}

File: spring-validate/src/main/kotlin/com/springvalidate/controller/ApiController.kt
package com.springvalidate.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.springvalidate.dto.PersonRequest
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import javax.validation.Valid
import javax.validation.constraints.Max
import javax.validation.constraints.Min


@Validated
@RestController
@RequestMapping(value = ["/api"])
class ApiController {

//    https://fullstackcode.dev/2022/05/10/complete-guide-to-spring-boot-validation/

    private val logger = LoggerFactory.getLogger(this.javaClass)

    @Autowired
    @Qualifier("customObjectMapper")
    lateinit var mapper: ObjectMapper

    @PostMapping("/createPerson")
    fun createPerson(@Valid @RequestBody request: PersonRequest) {

    }

    @PostMapping("/dept/{empId}/{depthId}")
    fun updateEmpDepartment(
        @PathVariable("empId") @Min(10, message = "empId >= 10") @Max(100, message = "empId <= 100")
        emp_id: Int?,

        @PathVariable("depthId") @Min(1001, message = "depthId >= 1001")
        depthId: Int?
    ) {

    }

//    @PostMapping("/createPersonBinding")
//    fun createPerson(@Valid @RequestBody request: PersonRequest, bindingResult: BindingResult): Any? {
//        if(bindingResult.hasErrors()) {
//            return bindingResult.getFieldErrors().associateBy({it -> (it as FieldError).field}, {it.defaultMessage})
//        }
//        return null
//    }

}